ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java

```diff
@@ -488,91 +523,106 @@ private void validateState(HttpContent httpContent)
+  private void continueReadIfPossible(long delta) {
+    if (!channel.config().isAutoRead()) {
+      if (bytesBuffered.addAndGet(delta) < bufferWatermark) {
+        channel.read();
+      } else {
+        nettyMetrics.watermarkOverflowCount.inc();
+      }
+    }
   }
 
   /**
-   * Decreases reference counts of content if required, updates the number of bytes read and invokes
-   * {@link ReadIntoCallbackWrapper#invokeCallback(Exception)} if {@code exception} is not {@code null} or if this is
-   * the last piece of content in the request.
-   * @param result The result of the request. This would be non null when the request executed successfully
-   * @param exception The exception that was reported on execution of the request
+   * Callback for each write into the given {@link AsyncWritableChannel}.
    */
-  @Override
-  public void onCompletion(Long result, Exception exception) {
-    if (httpContent != null) {
-      ReferenceCountUtil.release(httpContent);
+  protected class ContentWriteCallback implements Callback<Long> {
+    private final HttpContent httpContent;
+    private final boolean isLast;
+    private final ReadIntoCallbackWrapper callbackWrapper;
+
+    /**
+     * Creates a new instance of ContentWriteCallback.
+     * @param httpContent the {@link HttpContent} whose bytes were just written. Should be null if the data from the
+     *                    original {@link HttpContent} was copied and not "retained".
+     * @param isLast if this is the last piece of {@link HttpContent} for this request.
+     * @param callbackWrapper the {@link ReadIntoCallbackWrapper} that will receive updates of bytes read and one that
+     *                        should be invoked in {@link #onCompletion(Long, Exception)} if {@code isLast} is
+     *                        {@code true} or exception passed is not null.
+     */
+    public ContentWriteCallback(HttpContent httpContent, boolean isLast, ReadIntoCallbackWrapper callbackWrapper) {
+      this.httpContent = httpContent;
+      this.isLast = isLast;
+      this.callbackWrapper = callbackWrapper;
     }
-    callbackWrapper.updateBytesRead(result);
-    if (exception != null || isLast) {
-      callbackWrapper.invokeCallback(exception);
+
+    /**
+     * Decreases reference counts of content if required, updates the number of bytes read and invokes
+     * {@link ReadIntoCallbackWrapper#invokeCallback(Exception)} if {@code exception} is not {@code null} or if this is
+     * the last piece of content in the request.
+     * @param result The result of the request. This would be non null when the request executed successfully
+     * @param exception The exception that was reported on execution of the request
+     */
+    @Override
+    public void onCompletion(Long result, Exception exception) {
+      if (httpContent != null) {
+        ReferenceCountUtil.release(httpContent);
+      }
+      callbackWrapper.updateBytesRead(result);
+      continueReadIfPossible(-result);
```

Comment:
I assume this call to `continueReadIfPossible()` and its implementation are the only real changes here?